### PR TITLE
stats: refine the pseudo condition for stats

### PR DESCRIPTION
**Pull request by fipped**
This PR will use the average error rate of the row count estimated by stats to determine whether a column is pseudo. And when a column is not pseudo while the table is pseudo, we can use the stat of this column. The error rate information will be cleared when we analyzing the table.
PTAL @lamxTyler @coocood

### CODE REVIEW

**Comment by shenli** [comment]
@fipped Please fix the CI.

File: statistics/feedback.go [hunk 1]
```diff
@@ -554,28 +554,27 @@ func decodeFeedback(val []byte, q *QueryFeedback, c *CMSketch) error {
 		return errors.Trace(err)
 	}
 	// decode feedback for index
-	if len(pb.IndexRanges) > 0 {
+	if len(pb.Counts) > 0 {
```
Comment (lamxTyler) [inline statistics/feedback.go]: No need to test it.

File: statistics/feedback.go [hunk 2]
```diff
@@ -554,28 +554,27 @@ func decodeFeedback(val []byte, q *QueryFeedback, c *CMSketch) error {
...
 		}
-		return nil
-	}
-	// decode feedback for primary key
-	for i := 0; i < len(pb.IntRanges); i += 2 {
-		lower, upper := types.NewIntDatum(pb.IntRanges[i]), types.NewIntDatum(pb.IntRanges[i+1])
-		q.feedback = append(q.feedback, feedback{&lower, &upper, pb.Counts[i/2], 0})
 	}
 	return nil
+
```
Comment (lamxTyler) [inline statistics/feedback.go]: redundant blank line.

File: statistics/histogram.go [hunk 3]
```diff
@@ -635,12 +635,43 @@ func MergeHistograms(sc *stmtctx.StatementContext, lh *Histogram, rh *Histogram,
...
+	QueryTotal int64
+}
+
+// MaxErrorRate is the max error rate of estimate row count of a not pseudo column.
+// If the table is pseudo, but the average error rate is less than MaxErrorRate,
+// then the column is not pseudo.
+const MaxErrorRate = 0.25
+
+// IsPseudo is true when the total of query is zero or the average error
+// rate is less than MaxErrorRate.
```
Comment (lamxTyler) [inline statistics/histogram.go]: greater than.

File: statistics/table.go [hunk 4]
```diff
@@ -195,6 +207,8 @@ func (h *Handle) tableStatsFromStorage(tableInfo *model.TableInfo, loadAll bool)
 		// We copy it before writing to avoid race.
 		table = table.copy()
 	}
+	// After loading from storage, table stats becomes not pseudo.
+	table.Pseudo = false
```
Comment (lamxTyler) [inline statistics/table.go]: Why?

File: statistics/update.go [hunk 5]
```diff
@@ -50,6 +53,35 @@ func (m tableDeltaMap) update(id int64, delta int64, count int64, colSize *map[i
 	m[id] = item
 }
 
+type errorRateDelta struct {
+	ColErrorRate map[int64]*ErrorRate
```
Comment (lamxTyler) [inline statistics/update.go]: There is at most one col error rate which is the primary key.

File: statistics/update.go [hunk 6]
```diff
@@ -98,31 +130,84 @@ func mergeQueryFeedback(lq []*QueryFeedback, rq []*QueryFeedback) []*QueryFeedba
...
+	if !ok {
+		return nil
+	}
+	t := h.GetTableStats(table.Meta())
+	sc := h.ctx.GetSessionVars().StmtCtx
+
+	t = &(*t)
+	if t.Pseudo == true {
+		t.Pseudo = false
+		defer func() { t.Pseudo = true }()
```
Comment (lamxTyler) [inline statistics/update.go]: No need to reset it.

File: statistics/update.go [hunk 7]
```diff
@@ -98,31 +130,84 @@ func mergeQueryFeedback(lq []*QueryFeedback, rq []*QueryFeedback) []*QueryFeedba
...
+			ranges = append(ranges, &(ranger.Range{
+				LowVal:  []types.Datum{low},
+				HighVal: []types.Datum{high},
+			}))
+		}
+		var err error
+		if isIndex {
+			expected, err = t.GetRowCountByIndexRanges(sc, q.hist.ID, ranges)
+		} else {
+			if mysql.HasPriKeyFlag(t.Columns[q.hist.ID].Info.Flag) {
```
Comment (lamxTyler) [inline statistics/update.go]: It is always true here.

File: statistics/table.go (hunk 4 above)
Comment (fipped) [reply statistics/table.go]: When we use `planBuilder.getStatsTable()` to get the stats table, we will compute whether the table is pseudo. We don't need to compute it at here, therefore we set the Pseudo of the table false here.

File: executor/analyze.go [hunk 8]
```diff
@@ -229,6 +228,7 @@ func (e *AnalyzeIndexExec) buildStats() (hist *statistics.Histogram, cms *statis
 		}
 	}
 	hist.ID = e.idxInfo.ID
+	hist.IsAnalyzed = 1
```
Comment (lamxTyler) [inline executor/analyze.go]: No need to record it in the histogram.

File: session/session.go [hunk 9]
```diff
@@ -220,7 +220,14 @@ func (s *session) GetSessionManager() util.SessionManager {
 
 func (s *session) StoreQueryFeedback(feedback interface{}) {
 	if s.statsCollector != nil {
-		s.statsCollector.StoreQueryFeedback(feedback)
+		do, err := GetDomain(s.store)
+		if err != nil {
+			log.Warnf("domain not found.")
```
Comment (lamxTyler) [inline session/session.go]: Use `Debug` level.

File: session/session.go [hunk 10]
```diff
@@ -220,7 +220,14 @@ func (s *session) GetSessionManager() util.SessionManager {
...
+		}
+		err = s.statsCollector.StoreQueryFeedback(feedback, do.StatsHandle(), do.InfoSchema())
```
Comment (lamxTyler) [inline session/session.go]: If there is an error when getting the domain, it will panic here.

File: session/session.go [hunk 11]
```diff
@@ -220,7 +220,14 @@ func (s *session) GetSessionManager() util.SessionManager {
...
+		if err != nil {
+			log.Warnf("store query feedback error: ", err)
```
Comment (lamxTyler) [inline session/session.go]: Ditto.

File: statistics/update.go [hunk 12]
```diff
@@ -98,31 +139,78 @@ func mergeQueryFeedback(lq []*QueryFeedback, rq []*QueryFeedback) []*QueryFeedba
...
+	isIndex := q.hist.tp.Tp == mysql.TypeBlob
 
+	table, ok := is.TableByID(q.tableID)
+	if !ok {
+		return nil
+	}
+	t := h.GetTableStats(table.Meta())
+	sc := h.ctx.GetSessionVars().StmtCtx
+	if t.Pseudo == true {
+		ranges := make([]*ranger.Range, 0, len(q.feedback))
```
Comment (lamxTyler) [inline statistics/update.go]: Extract a function for it.

File: statistics/update.go [hunk 13]
```diff
@@ -98,31 +136,63 @@ func mergeQueryFeedback(lq []*QueryFeedback, rq []*QueryFeedback) []*QueryFeedba
...
 	} else {
-		rate = math.Abs(float64(q.expected-q.actual) / float64(q.actual))
+		rate = math.Abs(expected-float64(q.actual)) / float64(q.actual)
 	}
 	metrics.StatsInaccuracyRate.Observe(rate)
 
+	if h.rateMap == nil {
+		h.rateMap = make(map[int64]errorRateDelta)
+	}
+	h.rateMap.update(q.tableID, q.hist.ID, rate, isIndex)
```
Comment (coocood) [inline statistics/update.go]: This is map access is not protected by lock.

File: statistics/update.go [hunk 14]
```diff
@@ -298,6 +391,13 @@ func (h *Handle) HandleUpdateStats(is infoschema.InfoSchema) error {
 		}
 		// dump the stats into kv
 		if hist != nil {
+			// Update the NDV of primary key column.
+			if col != nil && mysql.HasPriKeyFlag(col.Info.Flag) {
```
Comment (coocood) [inline statistics/update.go]: Need to ensure table `PKIsHandle` is true?

File: statistics/update.go [hunk 15]
```diff
@@ -333,6 +433,12 @@ func (h *Handle) HandleUpdateStats(is infoschema.InfoSchema) error {
 			log.Debugf("decode feedback failed, err: %v", errors.ErrorStack(err))
 		}
 	}
+	// Update the NDV of primary key column.
+	if col != nil && mysql.HasPriKeyFlag(col.Info.Flag) {
```
Comment (coocood) [inline statistics/update.go]: ditto

File: statistics/feedback.go [hunk 16]
```diff
@@ -91,6 +91,30 @@ func (q *QueryFeedback) CollectFeedback(numOfRanges int) bool {
 	return true
 }
 
+// DecodeToRanges decode the feedback to ranges.
+func (q *QueryFeedback) DecodeToRanges(isIndex bool, isPK bool) ([]*ranger.Range, error) {
+	ranges := make([]*ranger.Range, 0, len(q.feedback))
+	for _, val := range q.feedback {
+		low, high := *val.lower, *val.upper
+		if !isIndex && isPK {
```
Comment (lamxTyler) [inline statistics/feedback.go]: `isIndex` and `isPK` cannot be the same, just one is enough.

File: statistics/histogram.go [hunk 17]
```diff
@@ -635,12 +635,43 @@ func MergeHistograms(sc *stmtctx.StatementContext, lh *Histogram, rh *Histogram,
 	return lh, nil
 }
 
+// ErrorRate is the error rate of estimate row count by bucket.
```
Comment (lamxTyler) [inline statistics/histogram.go]: `bucket` is not sufficient, it may also be cm sketch.

File: statistics/update.go [hunk 18]
```diff
@@ -274,6 +344,27 @@ func (h *Handle) dumpFeedbackToKV(fb *QueryFeedback) error {
...
+func (h *Handle) UpdateErrorRate(is infoschema.InfoSchema) {
+	tbls := make([]*Table, 0, len(h.rateMap))
+	for id, item := range h.rateMap {
+		table, ok := is.TableByID(id)
+		if !ok {
+			continue
+		}
+		tbl := h.GetTableStats(table.Meta())
+		if item.PkErrorRate != nil {
+			tbl.Columns[item.PkID].ErrorRate.merge(item.PkErrorRate)
```
Comment (lamxTyler) [inline statistics/update.go]: It may cause data race.

File: statistics/update.go [hunk 19]
```diff
@@ -274,6 +366,27 @@ func (h *Handle) dumpFeedbackToKV(fb *QueryFeedback) error {
...
+func (h *Handle) UpdateErrorRate(is infoschema.InfoSchema) {
+	tbls := make([]*Table, 0, len(h.rateMap))
+	for id, item := range h.rateMap {
+		table, ok := is.TableByID(id)
+		if !ok {
+			continue
+		}
+		tbl := h.GetTableStats(table.Meta()).copy()
+		if item.PkErrorRate != nil {
+			tbl.Columns[item.PkID].ErrorRate.merge(item.PkErrorRate)
```
Comment (lamxTyler) [inline statistics/update.go]: It may have data race.

File: statistics/update.go [hunk 20]
```diff
@@ -333,6 +457,12 @@ func (h *Handle) HandleUpdateStats(is infoschema.InfoSchema) error {
 			log.Debugf("decode feedback failed, err: %v", errors.ErrorStack(err))
 		}
 	}
+	// Update the NDV of primary key column.
+	if col != nil && mysql.HasPriKeyFlag(col.Info.Flag) && PKIsHandle {
+		if hist.NDV < int64(hist.totalRowCount()) {
+			hist.NDV = int64(hist.totalRowCount())
```
Comment (lamxTyler) [inline statistics/update.go]: It may have data race.

**Comment by lamxTyler** [comment]
PTAL @coocood

File: statistics/update_test.go [hunk 21]
```diff
@@ -371,6 +371,90 @@ func (s *testStatsUpdateSuite) TestAutoUpdate(c *C) {
...
+	}()
+	statistics.FeedbackProbability = 1
+
+	testKit := testkit.NewTestKit(c, s.store)
+	testKit.MustExec("use test")
+	testKit.MustExec("create table t (a bigint(64), b bigint(64), primary key(a), index idx(b))")
+	h.HandleDDLEvent(<-h.DDLEventCh())
+
+	for i := 1; i < 1001; i++ {
+		sql := fmt.Sprintf("insert into t values (%d, %d)", i*6+500, (i/20)*30+123)
```
Comment (coocood) [inline statistics/update_test.go]: Please add comments about this values to pick.

File: statistics/update_test.go [hunk 22]
```diff
@@ -371,6 +371,90 @@ func (s *testStatsUpdateSuite) TestAutoUpdate(c *C) {
...
+		testKit.MustExec(sql)
+	}
+
+	c.Assert(h.DumpStatsDeltaToKV(), IsNil)
+	testKit.MustExec("analyze table t")
+
+	for i := 0; i < 705; i++ {
+		sql := fmt.Sprintf("insert into t values (%d, %d)", i*2+501, (i/82)*120+500)
```
Comment (coocood) [inline statistics/update_test.go]: ditto

File: statistics/update_test.go [hunk 23]
```diff
@@ -371,6 +371,90 @@ func (s *testStatsUpdateSuite) TestAutoUpdate(c *C) {
...
+	table, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
+	tblInfo := table.Meta()
+
+	tbl := h.GetTableStats(tblInfo)
+	aID := tblInfo.Columns[0].ID
+	bID := tblInfo.Indices[0].ID
+
+	for i := 0; i < 30; i++ {
+		sql := fmt.Sprintf("select * from t where a between %d and %d", i*3+i/2, i*4+i/2+1200)
```
Comment (coocood) [inline statistics/update_test.go]: ditto

File: statistics/update_test.go [hunk 24]
```diff
@@ -371,6 +371,90 @@ func (s *testStatsUpdateSuite) TestAutoUpdate(c *C) {
...
+		tbl = h.GetTableStats(tblInfo)
+		checker := IsTrue
+		if i > 21 {
+			checker = IsFalse
+		}
+		c.Assert(tbl.Columns[aID].IsPseudo(), checker, Commentf("when `i` is %d, in sql `%s`", i, sql))
+	}
+
+	for i := 0; i < 15; i++ {
+		sql := fmt.Sprintf("select * from t where b between %d and %d", 2*i*i+555, 2*i*i+888)
```
Comment (coocood) [inline statistics/update_test.go]: ditto

File: statistics/update_test.go [hunk 25]
```diff
@@ -371,6 +371,90 @@ func (s *testStatsUpdateSuite) TestAutoUpdate(c *C) {
...
+		testKit.MustQuery(sql)
+		c.Assert(h.DumpStatsDeltaToKV(), IsNil)
+		c.Assert(h.DumpStatsFeedbackToKV(), IsNil)
+		c.Assert(h.HandleUpdateStats(is), IsNil)
+
+		h.UpdateErrorRate(is)
+		h.Update(is)
+		tbl = h.GetTableStats(tblInfo)
+		checker := IsTrue
+		if i == 0 || i > 9 {
```
Comment (coocood) [inline statistics/update_test.go]: Is it intentional to make error rate high when i == 0?

**Comment by shenli** [comment]
@coocood @lamxTyler PTAL

File: statistics/feedback.go [hunk 26]
```diff
@@ -91,6 +91,42 @@ func (q *QueryFeedback) CollectFeedback(numOfRanges int) bool {
...
 
+// DecodeToRanges decode the feedback to ranges.
+func (q *QueryFeedback) DecodeToRanges(isIndex bool) ([]*ranger.Range, error) {
+	ranges := make([]*ranger.Range, 0, len(q.feedback))
+	for _, val := range q.feedback {
+		low, high := *val.lower, *val.upper
+		var lowVal, highVal []types.Datum
+		if isIndex {
+			var err error
+			lowVal, err = codec.Decode(low.GetBytes(), low.Length())
```
Comment (lamxTyler) [inline statistics/feedback.go]: Why `Length`?

File: statistics/update.go [hunk 27]
```diff
@@ -98,31 +162,71 @@ func mergeQueryFeedback(lq []*QueryFeedback, rq []*QueryFeedback) []*QueryFeedba
...
 
+	table, ok := is.TableByID(q.tableID)
+	if !ok {
+		return nil
+	}
+	t := h.GetTableStats(table.Meta())
+	sc := h.ctx.GetSessionVars().StmtCtx
+
+	// A pseudo statistics table will be used when table row count from statistics is zero.
+	if t.Count == 0 {
```
Comment (lamxTyler) [inline statistics/update.go]: In this case, there is no need to go on, and it could not happen here.

File: statistics/update.go [hunk 28]
```diff
@@ -98,31 +162,71 @@ func mergeQueryFeedback(lq []*QueryFeedback, rq []*QueryFeedback) []*QueryFeedba
...
+		t = PseudoTable(table.Meta())
+	}
+
+	// The table be will pseudo when statistics is outdated.
+	if float64(t.ModifyCount)/float64(t.Count) > RatioOfPseudoEstimate {
+		tbl := *t
+		tbl.Pseudo = true
```
Comment (lamxTyler) [inline statistics/update.go]: It is better to set it when building the table struct.

File: statistics/update.go [hunk 29]
```diff
@@ -98,31 +162,71 @@ func mergeQueryFeedback(lq []*QueryFeedback, rq []*QueryFeedback) []*QueryFeedba
...
+		t = &tbl
+	}
+
+	if t.Pseudo == true {
```
Comment (lamxTyler) [inline statistics/update.go]: If the column stats is valid, there is no need to recalculate too.

**Comment by lamxTyler** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Pull request merged**